Add parse growth file task

guochen0 opened the pull request:
Resolves #13

Commit:
Initial edits on parse growth file

Commit:
Add preliminary code for parse growth file

Commit:
Update the python testing version

Commit:
Fix module not found error in tests

Commit:
Remove cache folder

Commit:
Adjust import orders

Commit:
Add extract file mocking with single member

Commit:
Add extract file mocking with single member

Commit:
Edit the test to include multiple extracted files and add docstrings

Commit:
Resolve merge conflicts

Commit:
Fix mypy error in state conversion function by returning None

Commit:
Complete test for parse growth file

guochen0 commented:
Hi @jessicasyu, I completed the `parse_growth_file` and the corresponding test. Please only review these two.

A few things I would like to mention:
1. I edited the` __init__` in `src/output/ `so I added some code for `parse_param_file.py` to prevent errors.  I think it's better if I complete `parse_param_file` after you review this round since it is very similar to `parse_growth_file`. The lint is not passing because `parse_param_file` is incomplete.
2. There are two functions: `convert_state_to_string` and `parse_growth_timepoint` that I use in `parse_growth_file`. I am not sure if the tests of those are needed.
3. I modified the tox.ini to accommodate my local testing environment. I will revert to the original python version later.

Thank you so much!

jessicasyu commented on src/arcade_collection/output/parse_growth_file.py:
```diff
@@ -0,0 +1,143 @@
+import tarfile
+from typing import List, Union
+
+import ntpath
```
No need to use `ntpath`, just use `os.path` instead, which will automatically pick the right module for the OS the code is running on.

jessicasyu commented on src/arcade_collection/output/parse_growth_file.py:
```diff
@@ -0,0 +1,143 @@
+    all_timepoints = []
+    for member in tar.getmembers():
+        extracted_member = tar.extractfile(member)
+        if extracted_member is not None:
```
I can't think of a reason the extracted member would be `None`

jessicasyu commented on src/arcade_collection/output/parse_growth_file.py:
```diff
@@ -0,0 +1,143 @@
+    for member in tar.getmembers():
+        extracted_member = tar.extractfile(member)
+        if extracted_member is not None:
+            base = ntpath.basename(member.name)
```
There should be no folders in the tar archive

jessicasyu commented on src/arcade_collection/output/parse_growth_file.py:
```diff
@@ -0,0 +1,143 @@
+            base_file = path.splitext(base)
+            file_name = base_file[0]
+            extension = base_file[1]
+            if file_name[0] != "." and extension == ".json":
```
Extension is always `.json`, so no need to check this

jessicasyu commented on src/arcade_collection/output/parse_growth_file.py:
```diff
@@ -0,0 +1,143 @@
+                extracted_json = json.loads(extracted_member.read().decode("utf-8"))
+                seed = extracted_json["seed"]
+
+                for timepoint in extracted_json["timepoints"]:
```
This block can be simplified into a list comprehension (see the `parse_cell_file.py` as an example)

jessicasyu commented on src/arcade_collection/output/parse_growth_file.py:
```diff
@@ -0,0 +1,143 @@
+    return timepoints_df
+
+
+def convert_state_to_string(state_index: int) -> Union[str, None]:
```
This can just be a list, since you have the indices. It would also make sense to be able to pass that list as a parameter, since other versions of ARCADE (e.g. CARCADE) have a slightly different mapping of cell state index to name.

jessicasyu commented on src/arcade_collection/output/parse_growth_file.py:
```diff
@@ -0,0 +1,143 @@
+    return None
+
+
+def parse_growth_timepoint(timepoint: dict, seed: int) -> List[list]:
```
Don't mix `List` and `list`, just use `list` for type hints as of Python 3.9

jessicasyu commented on src/arcade_collection/output/parse_growth_file.py:
```diff
@@ -0,0 +1,143 @@
+
+def parse_growth_timepoint(timepoint: dict, seed: int) -> List[list]:
+    """
+    Parse one timepoint of the simulation
```
Could be more descriptive in docstrings: what was the original data structure and what is the new data structure? What modifications are being made (e.g. averaging, rounding) to the data?

jessicasyu commented on src/arcade_collection/output/parse_growth_file.py:
```diff
@@ -0,0 +1,143 @@
+            position = cell[3]
+            volume = np.round(cell[4])
+            if len(cell[5]) == 0:
+                cycle = -1
```
I don't think we should keep the `-1` convention for empty cycles. Maybe `None`

jessicasyu commented on src/arcade_collection/output/parse_growth_file.py:
```diff
@@ -0,0 +1,143 @@
+            population = cell[1]
+            state = cell[2]
+            position = cell[3]
+            volume = np.round(cell[4])
```
I don't think we need to round the volume

jessicasyu commented on src/arcade_collection/output/parse_growth_file.py:
```diff
@@ -0,0 +1,143 @@
+                cycle = -1
+            else:
+                cycle = np.round(np.mean(cell[5]))
+            time = timepoint["time"]
```
The `time` is the same for all the cells in this timepoint, so this variable can be set outside the loop

jessicasyu commented on tests/arcade_collection/output/test_parse_growth_file.py:
```diff
@@ -0,0 +1,137 @@
+    def test_parse_growth_timepoint(self):
+        tar_object = mock.Mock(spec=tarfile.TarFile)
+        tar_object.name = "tar_object_name.tar"
+        assert tar_object.name == "tar_object_name.tar"
```
Why the extra `asserts`?

jessicasyu commented on tests/arcade_collection/output/test_parse_growth_file.py:
```diff
@@ -0,0 +1,137 @@
+class TestParseGrowthFile(unittest.TestCase):
+    def test_parse_growth_timepoint(self):
+        tar_object = mock.Mock(spec=tarfile.TarFile)
+        tar_object.name = "tar_object_name.tar"
```
Tar archives are compressed, so the extension is `tar.xz`

jessicasyu commented on tests/arcade_collection/output/test_parse_growth_file.py:
```diff
@@ -0,0 +1,137 @@
+        assert first_tar_member.name == "first_member.json"
+
+        second_tar_member = mock.Mock(spec=tarfile.TarInfo)
+        second_tar_member.name = "second_member.txt"
```
The model does not output `.txt` files

jessicasyu reviewed (changes requested):
Good start! The structure of the repo has changed so you'll need to fix merge conflicts. In particular, revisit how the `__init__.py` is handled.

Commit:
Update parse growth file and tests after initial review

Commit:
convert state to list format

jessicasyu commented on src/arcade_collection/output/__init__.py:
```diff
@@ -3,3 +3,5 @@
 from .parse_cells_file import parse_cells_file
 from .parse_locations_file import parse_locations_file
+from .parse_growth_file import parse_growth_file
+from .parse_params_file import parse_params_file
```
Make import order alphabetical

jessicasyu commented on src/arcade_collection/output/parse_growth_file.py:
```diff
@@ -0,0 +1,132 @@
+        u = int(location[0])
+        v = int(location[1])
+        w = int(location[2])
+        z = int(location[3])
```
Can simplify this to:

```python
u, v, w, z = location
```

jessicasyu commented on src/arcade_collection/output/parse_growth_file.py:
```diff
@@ -0,0 +1,132 @@
+    return timepoints_df
+
+
+def convert_state_to_string(state_index: int, state_list: List[str]) -> Union[str, None]:
```
You don't need this method. Just have a variable like `CELL_STATES = ["UNDEFINED", "APOPTOTIC", "QUIESCENT", "MIGRATORY", "PROLIFERATIVE", "SENESCENT", "NECROTIC"]` and index it directly with the state in the `parse_growth_timepoint` method.

jessicasyu commented on src/arcade_collection/output/parse_growth_file.py:
```diff
@@ -0,0 +1,132 @@
+        z = int(location[3])
+
+        for cell in cells:
+            population = cell[1]
```
Can simplify to:

```
_, population, state, position, volume, cycles = cell
```

rather than getting each item from the list separately.

jessicasyu commented on src/arcade_collection/output/parse_growth_file.py:
```diff
@@ -0,0 +1,132 @@
+                z,
+                position,
+                population,
+                convert_state_to_string(state, ["NEU", "APO", "QUI", "MIG", "PRO", "SEN", "NEC"]),
```
Use the full names for cell states

jessicasyu reviewed (changes requested):
@Isabelle-C This version is pretty far behind main at this point, I would merge main into this branch and fix any merge conflicts/linting errors (I think there are a couple mypy issues). It might make sense to have this PR just be for the parse growth file task, and put the parse params file task into a different PR.

Commit:
Merge branch 'main' into isabellec/arcadev2output

Commit:
Update according to feedback comments

Commit:
Add isort changes

Commit:
Attempt to fix mypy error

guochen0 commented:
@jessicasyu Sounds great! I merged main, fixed and mypy error, and deleted the parse param. Once this PR is finished, I will open a new one for the parse param.

I was wondering in the `src/arcade_collection/output/__init__.py`, there is a block with things like `parse_cells_file = task(parse_cells_file)`. I tried to do the same for parse growth, but the pytest does not work. There is a `TypeError: module, class, method, function, traceback, frame, or code object was expected, got Task for pytest`. I did attempt to resolve this, but couldn't. Thank you!

jessicasyu committed:
Merge branch 'main' into isabellec/arcadev2output

jessicasyu committed:
Remove tox.ini file

jessicasyu committed:
Move task decorator to init

jessicasyu committed:
Update docstrings

jessicasyu committed:
Remove fn call in test
